[Advanced Paste] Repeatedly pasting plain text as JSON produces inconsistent results

GhostVaibhav opened the pull request:
## Summary of the Pull Request
- This PR fixes the overwriting of the formatted text again and again.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #33944 
- [x] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [x] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [x] **Dev docs:** Added/updated
- [x] **New binaries:** Added on the required places
   - [x] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [x] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [x] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [x] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [x] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
- This PR basically keeps the original string in the clipboard, rather than copying and pasting the formatted values.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
- Tested it manually

GhostVaibhav commented:
@htcfreek
I've made some changes to the files and opened a draft PR. Could you have a look at it since the code is not working? The result that I'm getting now is - 

- Copied text - `something,that,i,know,about`
- Formatted text ("Paste as JSON") - `something,that,i,know,about`

It doesn't do anything. Could you help me with it?

htcfreek commented:
There is this `SendPasteKeyCombination` setting. If it is disabled your code can't work as you reset the clipboard before pasting.

GhostVaibhav commented:
@htcfreek 
I think the issue is something else. Because even if the setting is `false`, the output still doesn't align. Let me explain -

- When we press the "Paste to JSON option", the function is triggered.
- The original value is copied into the `tempText` variable. Also, since it's just a `get` function, I guess there is no modification in the clipboard data itself.
- The same clipboard data is passed to the designated function and the result is stored in the `jsonText` variable.
- Now, we set the `jsonText` to the clipboard data.
- Now, let's have the benefit of the doubt, and say the method inside _if_ is not executed.
- We set the `tempText` to the clipboard data.

That doesn't explain why it pastes the original data. It shouldn't, right? Or that I think. I would love to hear from you regarding this.

GhostVaibhav commented:
Hi @htcfreek, I added some log statements in between to see what was stored in different locations. And, I have attached the resulting image of the logs.

```c#
internal void ToJsonFunction(bool pasteAlways = false)
{
    try
    {
        Logger.LogTrace();
        string tempText = ClipboardData.GetTextAsync().GetAwaiter().GetResult();
        Logger.LogInfo("Temp text: " + tempText);
        
        string jsonText = JsonHelper.ToJsonFromXmlOrCsv(ClipboardData);
        Logger.LogInfo("JSON text: " + jsonText);

        SetClipboardContentAndHideWindow(jsonText);
        Logger.LogInfo("Copied JSON text to the clipboard");

        if (pasteAlways || _userSettings.SendPasteKeyCombination)
        {
            Logger.LogInfo("Clipboard data: " + ClipboardData.GetTextAsync().GetAwaiter().GetResult());

            Logger.LogInfo("Inside paste if");
            ClipboardHelper.SendPasteKeyCombination();
            Logger.LogInfo("End paste if");
        }

        SetClipboardContentAndHideWindow(tempText);
        string temp = ClipboardData.GetTextAsync().GetAwaiter().GetResult();
        Logger.LogInfo("Text in clipboard: " + temp);
    }
    catch
    {
    }
}
```

![image](https://github.com/user-attachments/assets/d79582e8-a847-41b8-a4de-87cd421930b0)

One question, that I was having, how come the clipboard content on paste changes to some other value or am I missing something?

htcfreek commented:
@GhostVaibhav
I am building a debug build now. Then I can investigate things tomorrow. _(One first idea I have regarding your logging is, that the logger might not live long enough after closing the window.)_

htcfreek commented:
@GhostVaibhav 
After doing some investigation I can tell you the following:
1. Your code works great (see logfile) except of a small timing problem: Executing the paste key needs to much time and clipboard gets reset before the paste action was executed.
   - You need to add a `thread.Sleep()` between sending the paste key and resetting the clipboard in `OptionsViewModel::ToJsonFunction`. I have succesfully tested it with 250 ms. 
2. That you don't see the last log line "Text in clipboard:" with your [modified code](https://github.com/microsoft/PowerToys/pull/34151#issuecomment-2286690218) can have two reasons:
   1. A crash in `ClipboardHelper::SetClipboardTextContent` that is not catched in a try-catch block.
   2. The clipboard is still accessed/locked by the set clipboard command and your code line `string temp = ClipboardData.GetTextAsync().GetAwaiter().GetResult();` in `ToJsonFunction` method crashes.
3. There is a big timing problem in `ClipboardHelper::SetClipboardTextContent` which causes `Clipboard.Flush()` to mostly not work.
   - Can be fixed by adding a while loop that waits until the clipboard update has finished.



## Logs and investigation informations
_Code changes are based on your public PR code._

### Changes on `OptionsViewModel::ToJsonFunction`
- Added thread.Sleep command.
- Added two comments.
- Added log to know if the end of the method was reached.

```csharp
	    ClipboardHelper.SendPasteKeyCombination();
	}

	// Sleeping 250 ms that paste command can be completely executed before resetting the clipboard.
	// Otherwise we reset the clipboard to early and past the original content. Or the clipboard is still used while resetting and restting fails.
	Thread.Sleep(250);

	// Resetting clipboard content
	SetClipboardContentAndHideWindow(tempText);
	Logger.LogInfo("End of method 'ToJsonFunction' reached.");
```

### Changes on `ClipboardHelper::SetClipboardTextContent`
- Added logging for the current clipboard content and the content to set. (For easier debugging.)
- Added logging if the set clipboard command returned true or false.
- Improved logging of errors for `Clipboard.Flush()` by adding a try-catch inside the Task command.
- Added a while loop to wait until the clipboard text is set completely.
```csharp
internal static void SetClipboardTextContent(string text)
{
    Logger.LogTrace();

    if (!string.IsNullOrEmpty(text))
    {
        Logger.LogDebug("Text to set to clipboard: " + text);

        DataPackage output = new();
        output.SetText(text);
        bool success = Clipboard.SetContentWithOptions(output, null);
        Logger.LogInfo("Setting Clipboard data was success? : " + success);

        // Wait 50 ms in a loop until setting the clipboard content has really finished.
        while (!Clipboard.GetContent().Contains("Text"))
        {
            Thread.Sleep(50);
        }

        try
        {
            string clipContent = Clipboard.GetContent().GetTextAsync().GetAwaiter().GetResult();
            Logger.LogInfo("Clipboard content for current process:" + clipContent);
        }
        catch (Exception ex)
        {
            Logger.LogError("Falied to get clipboard content: ", ex.GetBaseException());
        }

        // TODO(stefan): For some reason Flush() fails from time to time when directly activated via hotkey.
        // Calling inside a loop makes it work.
        bool flushed = false;
        for (int i = 0; i < 5; i++)
        {
            if (flushed)
            {
                break;
            }

            try
            {
                Task.Run(() =>
                {
                    try
                    {
                        Clipboard.Flush();
                    }
                    catch (Exception ex)
                    {
                        Logger.LogError("Clipboard.Flush() failed. Real reason:", ex.GetBaseException());
                        throw;
                    }
                }).Wait();

                flushed = true;
                Logger.LogDebug("Clipboard flushed.");
            }
            catch (Exception ex)
            {
                Logger.LogError("Clipboard.Flush() failed", ex);
            }
        }
    }
}
```

### Log ouput
- Note: The originally copied text was `abcdef dsa dsaldsa`.
```
[12:49:06.5536638] [Debug] JsonHelper::ToJsonFromXmlOrCsv
    Convert from plain text.
[12:49:06.5537349] [Trace] ClipboardHelper::SetClipboardTextContent
[12:49:06.5537628] [Debug] ClipboardHelper::SetClipboardTextContent
    Text to set to clipboard: [
  "abcdef dsa dsaldsa"
]
[12:49:06.5542858] [Info] ClipboardHelper::SetClipboardTextContent
    Setting Clipboard data was success? : True
[12:49:06.5838881] [Info] ClipboardHelper::SetClipboardTextContent
    Clipboard content for current process:[
  "abcdef dsa dsaldsa"
]
[12:49:06.6040774] [Debug] ClipboardHelper::SetClipboardTextContent
    Clipboard flushed.
[12:49:06.7492502] [Trace] ClipboardHelper::SendPasteKeyCombination
[12:49:06.7908409] [Info] ClipboardHelper::SendPasteKeyCombination
    Paste sent
[12:49:06.7912791] [Trace] ClipboardHelper::SetClipboardTextContent
[12:49:06.7913137] [Debug] ClipboardHelper::SetClipboardTextContent
    Text to set to clipboard: abcdef dsa dsaldsa
[12:49:06.8134983] [Info] ClipboardHelper::SetClipboardTextContent
    Setting Clipboard data was success? : True
[12:49:06.9234363] [Info] ClipboardHelper::SetClipboardTextContent
    Clipboard content for current process:abcdef dsa dsaldsa
[12:49:06.9561999] [Debug] ClipboardHelper::SetClipboardTextContent
    Clipboard flushed.
[12:49:06.9986701] [Info] OptionsViewModel::ToJsonFunction
    End of method 'ToJsonFunction' reached.
```

---

@stefansjfw
1. Should I open an new issue and PR to implement the loop inside all set clipboard methods to fix the flush problem?
2. Do you have a better idea than using `Thread.Sleep()` for waiting until the paste action is finally executed?

GhostVaibhav commented:
Hi @htcfreek, I re-implemented both files as per your suggestions, but they are still not working for me.

So now, when I copy the text and paste the JSON. The correct JSON is pasted, but the clipboard content is not updated. The clipboard contains the pasted JSON itself.

Also, I thought that this could be a timing issue, so I increased the thread's sleep time from `250` to `500`, but it is still not working.

htcfreek commented:
@GhostVaibhav 
Can you share the logging output?

GhostVaibhav commented:
The logging output is huge. Therefore, I have attached the whole log file. Fyi, it's filled with flush errors.

[Log_2024-09-01.txt](https://github.com/user-attachments/files/16828711/Log_2024-09-01.txt)

htcfreek commented:
@GhostVaibhav 
Wow that's strange: The clipboard data gets updated. But when flushing the clipboard to make the content accessible by other apps it throws a strange error:
```
Clipboard.Flush() failed. Real reason:
Unspecified error

The operation is not permitted because the calling application is not the owner of the data on the clipboard.
Inner exception: 
```

htcfreek commented:
@GhostVaibhav 
After finding [this article](https://learn.microsoft.com/en-us/dotnet/api/system.windows.clipboard.iscurrent?view=windowsdesktop-8.0#system-windows-clipboard-iscurrent(system-windows-idataobject)) I am wondering if we can do the following two things:
1. Instead of setting and then flushing using `Clipboard.SetDataObject(data, true)`.
2. After 1 validating the content with `bool isOriginalDataObject = Clipboard.IsCurrent(data);` instead the way I suggested. Because comparing the whole object validates data and data type.

GhostVaibhav commented:
Hi @htcfreek, it now works. I made some changes to your code. 

1. The most important change, I added was a sleep of 50 ms before doing another `Clipboard.Flush()`.
```cs
try
{
    Task.Run(() =>
    {
        try
        {
            Clipboard.Flush();
        }
        catch (Exception ex)
        {
            Logger.LogError("Clipboard.Flush() failed. Real reason:", ex.GetBaseException());
            throw;
        }
    }).Wait();

    flushed = true;
    Logger.LogDebug("Clipboard flushed.");
}
catch (Exception ex)
{
    Logger.LogError("Clipboard.Flush() failed", ex);
    Thread.Sleep(50); // This
}
```
2. Also, you forgot to use the variable in the checking of the clipboard getting updated, so I changed that as well.
```cs
// Wait 50 ms in a loop until setting the clipboard content has finished.
while (!Clipboard.GetContent().GetTextAsync().GetAwaiter().GetResult().Contains(text))
{
    Thread.Sleep(50);
}
```

But there are still exceptions about the flush in the log file. Should that be of any concern now?

htcfreek commented:
@GhostVaibhav 
If it now works I am fine with it.
Do we need the adjustments for plain text and markdown too?

GhostVaibhav commented:
The changes are done in `ClipboardHelper.cs` which is then used in all three functions, so, I think, the answer is a yes.

htcfreek commented:
@GhostVaibhav 
Can you commit and push please.

htcfreek commented on src/modules/AdvancedPaste/AdvancedPaste/Helpers/ClipboardHelper.cs:
```diff
@@ -21,9 +20,28 @@ internal static void SetClipboardTextContent(string text)
+                try
+                {
+                    string clipContent = Clipboard.GetContent().GetTextAsync().GetAwaiter().GetResult();
+                    Logger.LogInfo("Clipboard content for current process:" + clipContent);
```
let us use debug log. Thoughts?

And I missed a space after the ":" to separate text and clipboard content.

GhostVaibhav replied:
I think, yes, we should go ahead and use debug log here.

GhostVaibhav replied:
Do you think we should replace it below as well?
```cs
DataPackage output = new();
output.SetText(text);
bool success = Clipboard.SetContentWithOptions(output, null);
Logger.LogInfo("Setting Clipboard data was success? : " + success);
```

GhostVaibhav replied:
I have replaced both of them in the latest commit. Let me know if we need any more changes.

htcfreek replied:
As we have no explicit error log it's better to use info log for the "set clipboard" result. We don't log debug level on the machines of our users.

I suggested the debug log for the clipboard content to set for two reasons:
1. Privacy. We should not log every clipboard data especially without errors.
2. Keep log file smaller.

GhostVaibhav replied:
I have replaced them with Debug since that information is only needed when we're actually debugging the code. Also, we don't want to sneak peek at the clipboard content whenever the user calls the module.

htcfreek commented on src/modules/AdvancedPaste/AdvancedPaste/ViewModels/OptionsViewModel.cs:
```diff
@@ -291,15 +311,28 @@ internal void ToJsonFunction(bool pasteAlways = false)
             try
             {
                 Logger.LogTrace();
+                string tempText = ClipboardData.GetTextAsync().GetAwaiter().GetResult();
+                Logger.LogInfo("Temp text: " + tempText);
```
remove all the logging used for testing here. (obsolete as we log in clipboard helper.)

htcfreek commented on src/modules/AdvancedPaste/AdvancedPaste/ViewModels/OptionsViewModel.cs:
```diff
@@ -291,15 +311,28 @@ internal void ToJsonFunction(bool pasteAlways = false)
+                    Logger.LogInfo("End paste if");
                 }
+
+                Thread.Sleep(250);
+                SetClipboardContentAndHideWindow(tempText);
+                string temp = ClipboardData.GetTextAsync().GetAwaiter().GetResult();
+                Logger.LogInfo("Text in clipboard: " + temp);
```
please align with the code in the other methods and add the comments here too.

htcfreek commented on src/modules/AdvancedPaste/AdvancedPaste/ViewModels/OptionsViewModel.cs:
```diff
@@ -5,7 +5,9 @@
 using System;
 using System.Collections.ObjectModel;
 using System.Globalization;
+using System.Linq;
```
why do we need this using?

htcfreek commented on src/modules/AdvancedPaste/AdvancedPaste/ViewModels/OptionsViewModel.cs:
```diff
@@ -252,6 +254,8 @@ internal void ToPlainTextFunction()
             {
                 Logger.LogTrace();
 
+                string tempText = ClipboardData.GetTextAsync().GetAwaiter().GetResult();
```
for clarity please rename this. Maybe `originalClipboardData`.

htcfreek reviewed (commented):
Left some comments. Didn't tested again yet.

htcfreek commented on src/modules/AdvancedPaste/AdvancedPaste/Helpers/ClipboardHelper.cs:
```diff
@@ -21,9 +20,28 @@ internal static void SetClipboardTextContent(string text)
+                    Thread.Sleep(50);
+                }
+
+                try
+                {
+                    string clipContent = Clipboard.GetContent().GetTextAsync().GetAwaiter().GetResult();
+                    Logger.LogDebug("Clipboard content for current process: " + clipContent);
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogError("Failed to get clipboard content: ", ex.GetBaseException());
+                }
```
Can we mark this to only include in debug buid? 🤔 Logging an error for data we don't use in non-debug builds might be confusing.

GhostVaibhav replied:
Resolved in the latest commit

htcfreek commented:
Before you merge plese check this similar PR: #36404

shuaiyuanxx commented:
The root cause of this issue lies in continuously reading content from the clipboard and forcibly converting it into JSON. The correct and straightforward solution is to first check whether the content is already in the desired format, and only perform the conversion if it is not.

htcfreek commented:
Closing this as we have another solution merged yet.
